File: 234-palindrome-linked-list/palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if (head == null || head.next == null) 
            return true;
        
        ListNode fast = head;
        ListNode slow = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        if (fast != null)
            slow = slow.next;
            
        ListNode newHead = reverseLL(slow);
        
        ListNode first = head;
        ListNode second = newHead;

        while (second != null) {
            if (first.val != second.val) {
                reverseLL(newHead);
                return false;
            }

            first = first.next;
            second = second.next;
        }

        reverseLL(newHead);
        return true;
    }

    public ListNode reverseLL(ListNode head) {
        if (head == null || head.next == null) 
            return head;

        ListNode newHead = reverseLL(head.next);
        
        ListNode front = head.next;
        front.next = head;
        head.next = null;

        return newHead;
    }
}

File: 234-palindrome-linked-list/palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if head is None or head.next is None:
            return True
        
        fast, slow = head, head
        while fast is not None and fast.next is not None:
            slow = slow.next
            fast = fast.next.next
        
        if fast is not None:
            slow = slow.next

        newHead = self.reverseLL(slow)

        first, second = head, newHead
        while second is not None:
            if first.val != second.val:
                self.reverseLL(newHead)
                return False

            first = first.next
            second = second.next

        self.reverseLL(newHead)
        return True

    def reverseLL(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None:
            return head
        
        newHead = self.reverseLL(head.next)
        front = head.next
        front.next = head
        head.next = None

        return newHead
